sys/kern/link_elf.c

```diff
@@ -1790,8 +1790,17 @@ link_elf_symbol_values1(linker_file_t lf, c_linker_sym_t sym,
 			return (ENOENT);
 		symval->name = ef->strtab + es->st_name;
 		val = (caddr_t)ef->address + es->st_value;
-		if (ELF_ST_TYPE(es->st_info) == STT_GNU_IFUNC)
+		while (ELF_ST_TYPE(es->st_info) == STT_GNU_IFUNC) {
+			c_linker_sym_t sym1;
+			long off;
+
 			val = ((caddr_t (*)(void))val)();
+			if (link_elf_search_symbol(lf, (ptraddr_t)val, &sym1,
+			    &off) != 0 || off != 0)
+				return (ENOENT);
+			es = (const Elf_Sym *)sym1;
+			val = (caddr_t)ef->address + es->st_value;
+		}
 #ifdef __CHERI_PURE_CAPABILITY__
 		val = make_capability(es, val);
 #endif
```

Comment:
Is the problem not just that we should be doing make_capability prior to the IFUNC check, rather than after?

Reply:
I don't think so, or I don't understand your suggestion.

Consumers of this interface generally don't want to know about ifuncs. What we do for ifuncs is provide a function symbol that refers to the implementation selected by the resolver. Without this change, we were kind of doing that, but the symbol value would be a sentry capability and the size would be that of the ifunc symbol (which in practice is the size of the resolver function). My goal with this change is to eliminate those differences. I don't see how calling make_capability() first would help with that.

Reply:
The return value of the resolver is what we want to be giving, because it itself comes from some other non-IFUNC capability that the linker's properly constructed. The problem is that make_capability, for an IFUNC, thinks it's been given a capability that's ef->address plus some constant, but it's not, it's got the actual function pointer. On the flip side, we're calling the resolver via a capability that's ef->address plus some constant, so hasn't had its permissions restricted. IFUNCs are in essence "get a capability to the function as normal, but then call it to get the real capability", but we're failing to implement that here.

Reply:
Ok, I think I see, but addressing that won't fix my problem: the resolver returns a function pointer, which isn't what callers of `link_elf_each_function_nameval()` expect. I'd have to, at least, unseal the symbol value.

Reply:
But that's what it returns for normal STT_FUNC today, too?

Reply:
No, for STT_FUNC the returned value is a capability derived from `ef->address`. (And in fact the bounds are not clamped to the symbol size, which seems like a bug in itself, as `ef->address` covers the address space(!))

For IFUNCs, we are instead returning the function pointer returned by the resolver, hence my attempt to make things more consistent.

Reply:
Hm, I see what you mean. @bsdjhb wrote this code, I'm not sure quite what it's meant to be doing, though I do feel we should be calling the IFUNC with the right capability (i.e. the one that it would have for normal symbol resolution). What are the (transitive) uses of this? Should we really be returning a ptraddr_t instead and forcing the users that do need provenance, potentially quite permissive, to ask for it explicitly?

Reply:
The main consumer of this interface is linker_file_lookup_set(), which definitely requires capabilities, though I expect that linker sets generally don't contain function symbols. I don't think we can return a ptraddr_t here.

I don't quite follow your comment about calling the ifunc resolver with the "right" capability - from what I can, relocate_file1() passes `ef->address` as `relocbase` to `elf_reloc_internal()`, where it's used to derive the resolver capability, just as we do here. You said earlier that we should perhaps call `make_capability()` before invoking the resolver, which is makes sense to me, but it doesn't solve my problem.

Reply:
As in the capability we use for the call should have code capability permissions only (and the right bounds, but I believe we don't bound kernel code capabilities even on Morello?)

Reply:
To be clear, you are asking me to make that change in this PR?

Reply:
No, other than that if we're thinking carefully about and rewriting this code to be less weird it might be something that gets folded into the rewrite rather than preserving the oddity.

Reply:
(I need to go and understand exactly what these interfaces are being used for to be able to assess what the right thing to do about your issue is)

Reply:
I think Jess is right that make_capability should be before we invoke the IFUNC.

The right fix might be that we just need to unseal the returned capability from the resolver.  The kernel linker should have a capability for that to-hand.

Looking at this more, I wonder if make_capability( ) should in fact be sealing function capabilities in general, and then for the case of DTrace we need to explicitly unseal code pointers?

I'm going to do a little set of changes to the kernel linker that fixes make_capability (and also fixes link_elf_obj.c which uses an inlined but not correct version).  Konrad might also have changes to honor PT_CHERI_PCC which will fix the bounds on code pointers (which are indeed too wide today).

Reply:
I've merged my linker changes, so you should be able to drop this commit when rebasing.  However, with the changes in `dev`, you will now always get sealed sentry capabilities for function symbols.  DDB doesn't use symbol values directly (and ifunc resolvers always return sentries anyway), so we will need to figure out the right way to give DTrace "code" pointers instead of function pointers.  The simplest (but possibly grossest) fix is to have DTrace use a powerful capability internally to unseal sentries it gets back from the linker.  Alternatively, if we can fit it into the API DTrace uses, we can add a flag to return code pointers.  Possibly we could even make those capabilities read-only without X permissions?

Reply:
I guess one thing I'm less certain of is do you need to re-lookup the symbol of the underlying function returned by the ifunc resolver?  In theory there might not even be a symbol as a resolver might return pointers to local functions?  Is it because you need the size to know the bounds of where you might insert probes for FBT or some such?

Reply:
We may also need the code vs function pointer distinction for c18n trampolines in the kernel

Reply:
(Although in the c18n case code pointers are still sentries, just not wrapped, so it's not entirely the same thing, but the APIs will need to grow in some way to support that anyway)